compiler/rustc_mir_transform/src/scalar_deref_prop.rs

```diff
@@ -0,0 +1,432 @@
+        debug!(assignments = ?ssa.assignments[local]);
+        let Set1::One(LocationExtended::Plain(loc)) = ssa.assignments[local] else {
+            targets[local] = Value::Unknown;
+            continue;
+        };
+
+        // If this a mutable reference that we cannot fully replace, mark it as unknown.
+        if body.local_decls[local].ty.is_mutable_ptr() && !fully_replacable_locals.contains(local) {
+            targets[local] = Value::Unknown;
+            continue;
+        }
+
+        // `loc` must point to a direct assignment to `local`.
+        let Either::Left(stmt) = body.stmt_at(loc) else { bug!() };
+        let Some((_target, rvalue)) = stmt.kind.as_assign() else { bug!() };
+        assert_eq!(_target.as_local(), Some(local));
+
+        let rvalue = match rvalue {
+            // This is a copy, just use the value we have in store for the previous one.
+            // As we are visiting in `assignment_order`, ie. reverse postorder, `rhs` should
+            // have been visited before.
+            Rvalue::Use(Operand::Copy(place) | Operand::Move(place))
+            | Rvalue::CopyForDeref(place) => {
+                if let Some(rhs) = place.as_local() {
+                    let target = targets[rhs];
+                    if matches!(target, Value::Pointer(_)) {
+                        target
+                    } else if matches!(ssa.assignments[rhs], Set1::One(_)) {
+                        Value::Pointer(tcx.mk_place_deref(rhs.into()))
+                    } else {
+                        Value::Unknown
+                    }
+                } else {
+                    Value::Unknown
+                }
+            }
+            Rvalue::Ref(_, _, place) | Rvalue::AddressOf(_, place) => {
+                let mut place = *place;
+                if place.projection.first() == Some(&PlaceElem::Deref)
+                    && let Value::Pointer(target) = targets[place.local]
+                {
+                    place = target.project_deeper(&place.projection[1..], tcx);
+                }
+                assert_ne!(place.local, local);
+                if is_constant(place) { Value::Pointer(place) } else { Value::Unknown }
+            }
+            // We do not know what to do, so mark as not-a-pointer.
+            _ => Value::Unknown,
+        };
+
+        targets[local] = rvalue;
+    }
+
+    debug!(?targets);
+
+    targets
+}
+
+/// Compute the set of locals that can be fully replaced.
+fn fully_replacable_locals(ssa: &SsaLocals, copies: &IndexVec<Local, Local>) -> BitSet<Local> {
+    let mut replacable = BitSet::new_empty(ssa.direct_uses.len());
+
+    // First pass: for each local, whether its uses can be fully replaced.
+    for (local, &num_direct_uses) in ssa.direct_uses.iter_enumerated() {
+        if num_direct_uses == 0 {
+            replacable.insert(local);
+        }
+    }
+
+    // Consolidate to have a local iff all its copies are.
```

Comment:
Leave a comment that this works, exactly because all possible `head`s point to themselves.